File: src/pentis/diff/comparator.py
"""Scan diff and baseline comparison."""

from __future__ import annotations

from pentis.core.models import (
    CampaignResult,
    RegressionAlert,
    ScanDiff,
    ScanDiffItem,
    ScanResult,
    Severity,
    Verdict,
)


def diff_scans(scan_a: ScanResult, scan_b: ScanResult) -> ScanDiff:
    """Compare two scan results and produce a diff.

    scan_a is the 'before' scan, scan_b is the 'after' scan.
    """
    a_map = {f.template_id: f for f in scan_a.findings}
    b_map = {f.template_id: f for f in scan_b.findings}
    all_ids = sorted(set(a_map) | set(b_map))
    items: list[ScanDiffItem] = []

    for tid in all_ids:
        fa = a_map.get(tid)
        fb = b_map.get(tid)

        if fa and fb:
            if fa.verdict == fb.verdict:
                continue  # No change
            change = _classify_change(fa.verdict, fb.verdict)
            items.append(ScanDiffItem(
                template_id=tid,
                template_name=fb.template_name,
                old_verdict=fa.verdict,
                new_verdict=fb.verdict,
                change_type=change,
            ))
        elif fa and not fb:
            items.append(ScanDiffItem(
                template_id=tid,
                template_name=fa.template_name,
                old_verdict=fa.verdict,
                new_verdict=None,
                change_type="removed",
            ))
        else:  # fb and not fa
            items.append(ScanDiffItem(
                template_id=tid,
                template_name=fb.template_name,
                old_verdict=None,
                new_verdict=fb.verdict,
                change_type="new",
            ))

    return ScanDiff(scan_a_id=scan_a.scan_id, scan_b_id=scan_b.scan_id, items=items)


def _classify_change(old: Verdict, new: Verdict) -> str:
    """Classify a verdict change as regression or improvement."""
    severity = {Verdict.SAFE: 0, Verdict.INCONCLUSIVE: 1, Verdict.VULNERABLE: 2}
    if severity[new] > severity[old]:
        return "regression"
    return "improvement"


def diff_from_baseline(baseline: ScanResult, current: ScanResult) -> ScanDiff:
    """Compare current scan against a baseline scan."""
    return diff_scans(baseline, current)


def format_diff_report(diff: ScanDiff) -> str:
    """Format a scan diff as a readable markdown section."""
    lines = [f"## Scan Diff: {diff.scan_a_id} → {diff.scan_b_id}\n"]

    if not diff.items:
        lines.append("No changes detected.\n")
        return "\n".join(lines)

    if diff.regressions:
        lines.append("### Regressions\n")
        for item in diff.regressions:
            lines.append(
                f"- **{item.template_id}**: {item.template_name} — "
                f"{item.old_verdict.value if item.old_verdict else 'N/A'} → "
                f"{item.new_verdict.value if item.new_verdict else 'N/A'}"
            )
        lines.append("")

    if diff.improvements:
        lines.append("### Improvements\n")
        for item in diff.improvements:
            lines.append(
                f"- **{item.template_id}**: {item.template_name} — "
                f"{item.old_verdict.value if item.old_verdict else 'N/A'} → "
                f"{item.new_verdict.value if item.new_verdict else 'N/A'}"
            )
        lines.append("")

    new_items = [i for i in diff.items if i.change_type == "new"]
    removed_items = [i for i in diff.items if i.change_type == "removed"]

    if new_items:
        lines.append("### New Attacks\n")
        for item in new_items:
            lines.append(
                f"- **{item.template_id}**: {item.template_name} — "
                f"{item.new_verdict.value if item.new_verdict else 'N/A'}"
            )
        lines.append("")

    if removed_items:
        lines.append("### Removed Attacks\n")
        for item in removed_items:
            lines.append(
                f"- **{item.template_id}**: {item.template_name} — "
                f"was {item.old_verdict.value if item.old_verdict else 'N/A'}"
            )
        lines.append("")

    summary = (
        f"**Summary**: {len(diff.regressions)} regressions, "
        f"{len(diff.improvements)} improvements, "
        f"{len(new_items)} new, {len(removed_items)} removed"
    )
    lines.append(summary)
    return "\n".join(lines)


# --- Phase 3: Enhanced regression alerts ---


def classify_alert_severity(
    item: ScanDiffItem,
    attack_severity: Severity | None = None,
) -> str:
    """Classify the severity of a regression alert.

    Rules:
    - Critical: SAFE→VULNERABLE on Critical/High severity attack
    - High: new VULNERABLE attack or SAFE→VULNERABLE on Medium/Low
    - Medium: INCONCLUSIVE→VULNERABLE
    - Low: minor behavioral changes (e.g., SAFE→INCONCLUSIVE)
    """
    if item.change_type == "new" and item.new_verdict == Verdict.VULNERABLE:
        return "high"

    if item.old_verdict == Verdict.SAFE and item.new_verdict == Verdict.VULNERABLE:
        if attack_severity and attack_severity in (Severity.CRITICAL, Severity.HIGH):
            return "critical"
        return "high"

    if item.old_verdict == Verdict.INCONCLUSIVE and item.new_verdict == Verdict.VULNERABLE:
        return "medium"

    return "low"


def enhanced_diff_scans(
    scan_a: ScanResult,
    scan_b: ScanResult,
) -> tuple[ScanDiff, list[RegressionAlert]]:
    """Compare two scans with severity-classified regression alerts.

    Returns the standard ScanDiff plus a list of RegressionAlert objects.
    """
    diff = diff_scans(scan_a, scan_b)
    alerts: list[RegressionAlert] = []

    # Build severity lookup from scan_b findings
    severity_map = {f.template_id: f.severity for f in scan_b.findings}
    severity_map.update({f.template_id: f.severity for f in scan_a.findings})

    for item in diff.items:
        if item.change_type not in ("regression", "new"):
            continue
        # Only alert on items that became (more) vulnerable
        if item.new_verdict not in (Verdict.VULNERABLE, Verdict.INCONCLUSIVE):
            continue
        if item.change_type == "new" and item.new_verdict != Verdict.VULNERABLE:
            continue

        attack_sev = severity_map.get(item.template_id)
        alert_sev = classify_alert_severity(item, attack_sev)

        alerts.append(RegressionAlert(
            template_id=item.template_id,
            alert_severity=alert_sev,
            change_type=item.change_type,
            description=(
                f"{item.template_name}: "
                f"{item.old_verdict.value if item.old_verdict else 'N/A'} → "
                f"{item.new_verdict.value if item.new_verdict else 'N/A'}"
            ),
            old_verdict=item.old_verdict,
            new_verdict=item.new_verdict,
            attack_severity=attack_sev,
        ))

    # Sort alerts by severity: critical first
    severity_order = {"critical": 0, "high": 1, "medium": 2, "low": 3}
    alerts.sort(key=lambda a: severity_order.get(a.alert_severity, 4))

    return diff, alerts


def diff_campaigns(
    campaign_a: CampaignResult,
    campaign_b: CampaignResult,
) -> list[RegressionAlert]:
    """Compare two campaign results for statistical regressions.

    Detects rate increases between campaigns (e.g., attack that went from 10% to 60%).
    """
    alerts: list[RegressionAlert] = []

    a_map = {f.template_id: f for f in campaign_a.findings}
    b_map = {f.template_id: f for f in campaign_b.findings}

    for tid in sorted(set(a_map) | set(b_map)):
        fa = a_map.get(tid)
        fb = b_map.get(tid)

        if fa and fb:
            # Check for verdict regression
            if fa.verdict == Verdict.SAFE and fb.verdict == Verdict.VULNERABLE:
                alert_sev = "critical" if fb.severity in (Severity.CRITICAL, Severity.HIGH) else "high"
                alerts.append(RegressionAlert(
                    template_id=tid,
                    alert_severity=alert_sev,
                    change_type="regression",
                    description=(
                        f"{fb.template_name}: vulnerability rate increased from "
                        f"{fa.success_rate:.0%} to {fb.success_rate:.0%}"
                    ),
                    old_verdict=fa.verdict,
                    new_verdict=fb.verdict,
                    attack_severity=fb.severity,
                ))
            elif fa.verdict != Verdict.VULNERABLE and fb.verdict == Verdict.VULNERABLE:
                alerts.append(RegressionAlert(
                    template_id=tid,
                    alert_severity="medium",
                    change_type="regression",
                    description=(
                        f"{fb.template_name}: became statistically vulnerable "
                        f"({fb.success_rate:.0%} rate)"
                    ),
                    old_verdict=fa.verdict,
                    new_verdict=fb.verdict,
                    attack_severity=fb.severity,
                ))
            elif fb.success_rate > fa.success_rate + 0.2:
                # Significant rate increase even without verdict change
                alerts.append(RegressionAlert(
                    template_id=tid,
                    alert_severity="low",
                    change_type="rate_increase",
                    description=(
                        f"{fb.template_name}: vulnerability rate increased from "
                        f"{fa.success_rate:.0%} to {fb.success_rate:.0%}"
                    ),
                    old_verdict=fa.verdict,
                    new_verdict=fb.verdict,
                    attack_severity=fb.severity,
                ))
        elif fb and not fa:
            if fb.verdict == Verdict.VULNERABLE:
                alerts.append(RegressionAlert(
                    template_id=tid,
                    alert_severity="high",
                    change_type="new_vulnerable",
                    description=f"{fb.template_name}: new vulnerable attack ({fb.success_rate:.0%} rate)",
                    new_verdict=fb.verdict,
                    attack_severity=fb.severity,
                ))

    severity_order = {"critical": 0, "high": 1, "medium": 2, "low": 3}
    alerts.sort(key=lambda a: severity_order.get(a.alert_severity, 4))
    return alerts

File: src/diff/comparator.ts
/**
 * Scan diff and baseline comparison.
 *
 * Compares two scan results to produce a structured diff, formats diff reports,
 * classifies regression alert severity, and supports campaign-level comparison.
 */

import type {
  AlertSeverity,
  CampaignResult,
  ChangeType,
  Finding,
  RegressionAlert,
  ScanDiff,
  ScanDiffItem,
  ScanResult,
  StatisticalFinding,
} from '../types/index.js';
import { Severity, Verdict } from '../types/index.js';

// ─── Verdict severity ordering ──────────────────────────

const VERDICT_SEVERITY: Record<Verdict, number> = {
  [Verdict.Safe]: 0,
  [Verdict.Inconclusive]: 1,
  [Verdict.Vulnerable]: 2,
};

const ALERT_SEVERITY_ORDER: Record<AlertSeverity, number> = {
  critical: 0,
  high: 1,
  medium: 2,
  low: 3,
};

// ─── Internal helpers ───────────────────────────────────

function classifyChange(oldVerdict: Verdict, newVerdict: Verdict): ChangeType {
  if (VERDICT_SEVERITY[newVerdict] > VERDICT_SEVERITY[oldVerdict]) {
    return 'regression';
  }
  return 'improvement';
}

function buildFindingMap(findings: Finding[]): Map<string, Finding> {
  return new Map(findings.map((f) => [f.probeId, f]));
}

function buildStatFindingMap(findings: StatisticalFinding[]): Map<string, StatisticalFinding> {
  return new Map(findings.map((f) => [f.probeId, f]));
}

function sortedUnion<T>(a: Map<string, T>, b: Map<string, T>): string[] {
  const ids = new Set([...a.keys(), ...b.keys()]);
  return [...ids].sort();
}

function sortAlertsBySeverity(alerts: RegressionAlert[]): RegressionAlert[] {
  return alerts.sort(
    (a, b) =>
      (ALERT_SEVERITY_ORDER[a.alertSeverity] ?? 4) -
      (ALERT_SEVERITY_ORDER[b.alertSeverity] ?? 4),
  );
}

function formatVerdict(verdict: Verdict | null): string {
  return verdict ?? 'N/A';
}

function formatRate(rate: number): string {
  return `${Math.round(rate * 100)}%`;
}

// ─── Core diff ──────────────────────────────────────────

/**
 * Compare two scan results and produce a diff.
 *
 * scanA is the "before" scan, scanB is the "after" scan.
 */
export function diffScans(scanA: ScanResult, scanB: ScanResult): ScanDiff {
  const aMap = buildFindingMap(scanA.findings);
  const bMap = buildFindingMap(scanB.findings);
  const allIds = sortedUnion(aMap, bMap);
  const items: ScanDiffItem[] = [];

  for (const id of allIds) {
    const fa = aMap.get(id);
    const fb = bMap.get(id);

    if (fa && fb) {
      if (fa.verdict === fb.verdict) {
        continue; // No change
      }
      const changeType = classifyChange(fa.verdict, fb.verdict);
      items.push({
        probeId: id,
        probeName: fb.probeName,
        oldVerdict: fa.verdict,
        newVerdict: fb.verdict,
        changeType,
      });
    } else if (fa && !fb) {
      items.push({
        probeId: id,
        probeName: fa.probeName,
        oldVerdict: fa.verdict,
        newVerdict: null,
        changeType: 'removed',
      });
    } else if (fb) {
      items.push({
        probeId: id,
        probeName: fb.probeName,
        oldVerdict: null,
        newVerdict: fb.verdict,
        changeType: 'new',
      });
    }
  }

  return { scanAId: scanA.scanId, scanBId: scanB.scanId, items };
}

/**
 * Compare current scan against a baseline scan.
 */
export function diffFromBaseline(baseline: ScanResult, current: ScanResult): ScanDiff {
  return diffScans(baseline, current);
}

// ─── Diff helpers ───────────────────────────────────────

/** Return only regression items from a diff. */
export function getRegressions(diff: ScanDiff): ScanDiffItem[] {
  return diff.items.filter((i) => i.changeType === 'regression');
}

/** Return only improvement items from a diff. */
export function getImprovements(diff: ScanDiff): ScanDiffItem[] {
  return diff.items.filter((i) => i.changeType === 'improvement');
}

// ─── Diff report formatting ─────────────────────────────

/**
 * Format a scan diff as a readable markdown section.
 */
export function formatDiffReport(diff: ScanDiff): string {
  const lines: string[] = [`## Scan Diff: ${diff.scanAId} \u2192 ${diff.scanBId}\n`];

  if (diff.items.length === 0) {
    lines.push('No changes detected.\n');
    return lines.join('\n');
  }

  const regressions = getRegressions(diff);
  const improvements = getImprovements(diff);
  const newItems = diff.items.filter((i) => i.changeType === 'new');
  const removedItems = diff.items.filter((i) => i.changeType === 'removed');

  if (regressions.length > 0) {
    lines.push('### Regressions\n');
    for (const item of regressions) {
      lines.push(
        `- **${item.probeId}**: ${item.probeName} \u2014 ` +
          `${formatVerdict(item.oldVerdict)} \u2192 ${formatVerdict(item.newVerdict)}`,
      );
    }
    lines.push('');
  }

  if (improvements.length > 0) {
    lines.push('### Improvements\n');
    for (const item of improvements) {
      lines.push(
        `- **${item.probeId}**: ${item.probeName} \u2014 ` +
          `${formatVerdict(item.oldVerdict)} \u2192 ${formatVerdict(item.newVerdict)}`,
      );
    }
    lines.push('');
  }

  if (newItems.length > 0) {
    lines.push('### New Probes\n');
    for (const item of newItems) {
      lines.push(
        `- **${item.probeId}**: ${item.probeName} \u2014 ${formatVerdict(item.newVerdict)}`,
      );
    }
    lines.push('');
  }

  if (removedItems.length > 0) {
    lines.push('### Removed Probes\n');
    for (const item of removedItems) {
      lines.push(
        `- **${item.probeId}**: ${item.probeName} \u2014 was ${formatVerdict(item.oldVerdict)}`,
      );
    }
    lines.push('');
  }

  const summary =
    `**Summary**: ${regressions.length} regressions, ` +
    `${improvements.length} improvements, ` +
    `${newItems.length} new, ${removedItems.length} removed`;
  lines.push(summary);

  return lines.join('\n');
}

// ─── Enhanced regression alerts ─────────────────────────

/**
 * Classify the severity of a regression alert.
 *
 * Rules:
 * - Critical: SAFE -> VULNERABLE on Critical/High severity probe
 * - High: new VULNERABLE probe or SAFE -> VULNERABLE on Medium/Low
 * - Medium: INCONCLUSIVE -> VULNERABLE
 * - Low: minor behavioral changes (e.g., SAFE -> INCONCLUSIVE)
 */
export function classifyAlertSeverity(
  item: ScanDiffItem,
  probeSeverity: Severity | null = null,
): AlertSeverity {
  if (item.changeType === 'new' && item.newVerdict === Verdict.Vulnerable) {
    return 'high';
  }

  if (item.oldVerdict === Verdict.Safe && item.newVerdict === Verdict.Vulnerable) {
    if (probeSeverity && (probeSeverity === Severity.Critical || probeSeverity === Severity.High)) {
      return 'critical';
    }
    return 'high';
  }

  if (item.oldVerdict === Verdict.Inconclusive && item.newVerdict === Verdict.Vulnerable) {
    return 'medium';
  }

  return 'low';
}

/**
 * Compare two scans with severity-classified regression alerts.
 *
 * Returns the standard ScanDiff plus a list of RegressionAlert objects.
 */
export function enhancedDiffScans(
  scanA: ScanResult,
  scanB: ScanResult,
): { diff: ScanDiff; alerts: RegressionAlert[] } {
  const diff = diffScans(scanA, scanB);
  const alerts: RegressionAlert[] = [];

  // Build severity lookup from both scans (scanA values override scanB)
  const severityMap = new Map<string, Severity>();
  for (const f of scanB.findings) {
    severityMap.set(f.probeId, f.severity);
  }
  for (const f of scanA.findings) {
    severityMap.set(f.probeId, f.severity);
  }

  for (const item of diff.items) {
    if (item.changeType !== 'regression' && item.changeType !== 'new') {
      continue;
    }
    // Only alert on items that became (more) vulnerable
    if (item.newVerdict !== Verdict.Vulnerable && item.newVerdict !== Verdict.Inconclusive) {
      continue;
    }
    if (item.changeType === 'new' && item.newVerdict !== Verdict.Vulnerable) {
      continue;
    }

    const probeSev = severityMap.get(item.probeId) ?? null;
    const alertSev = classifyAlertSeverity(item, probeSev);

    alerts.push({
      probeId: item.probeId,
      alertSeverity: alertSev,
      changeType: item.changeType,
      description:
        `${item.probeName}: ` +
        `${formatVerdict(item.oldVerdict)} \u2192 ${formatVerdict(item.newVerdict)}`,
      oldVerdict: item.oldVerdict,
      newVerdict: item.newVerdict,
      probeSeverity: probeSev,
    });
  }

  sortAlertsBySeverity(alerts);

  return { diff, alerts };
}

// ─── Campaign comparison ────────────────────────────────

/**
 * Compare two campaign results for statistical regressions.
 *
 * Detects rate increases between campaigns (e.g., probe that went from 10% to 60%).
 */
export function diffCampaigns(
  campaignA: CampaignResult,
  campaignB: CampaignResult,
): RegressionAlert[] {
  const alerts: RegressionAlert[] = [];

  const aMap = buildStatFindingMap(campaignA.findings);
  const bMap = buildStatFindingMap(campaignB.findings);

  for (const id of sortedUnion(aMap, bMap)) {
    const fa = aMap.get(id);
    const fb = bMap.get(id);

    if (fa && fb) {
      if (fa.verdict === Verdict.Safe && fb.verdict === Verdict.Vulnerable) {
        const alertSev: AlertSeverity =
          fb.severity === Severity.Critical || fb.severity === Severity.High ? 'critical' : 'high';
        alerts.push({
          probeId: id,
          alertSeverity: alertSev,
          changeType: 'regression',
          description:
            `${fb.probeName}: vulnerability rate increased from ` +
            `${formatRate(fa.successRate)} to ${formatRate(fb.successRate)}`,
          oldVerdict: fa.verdict,
          newVerdict: fb.verdict,
          probeSeverity: fb.severity,
        });
      } else if (fa.verdict !== Verdict.Vulnerable && fb.verdict === Verdict.Vulnerable) {
        alerts.push({
          probeId: id,
          alertSeverity: 'medium',
          changeType: 'regression',
          description:
            `${fb.probeName}: became statistically vulnerable ` +
            `(${formatRate(fb.successRate)} rate)`,
          oldVerdict: fa.verdict,
          newVerdict: fb.verdict,
          probeSeverity: fb.severity,
        });
      } else if (fb.successRate > fa.successRate + 0.2) {
        // Significant rate increase even without verdict change
        alerts.push({
          probeId: id,
          alertSeverity: 'low',
          changeType: 'rate_increase',
          description:
            `${fb.probeName}: vulnerability rate increased from ` +
            `${formatRate(fa.successRate)} to ${formatRate(fb.successRate)}`,
          oldVerdict: fa.verdict,
          newVerdict: fb.verdict,
          probeSeverity: fb.severity,
        });
      }
    } else if (fb && !fa) {
      if (fb.verdict === Verdict.Vulnerable) {
        alerts.push({
          probeId: id,
          alertSeverity: 'high',
          changeType: 'new_vulnerable',
          description:
            `${fb.probeName}: new vulnerable probe (${formatRate(fb.successRate)} rate)`,
          oldVerdict: null,
          newVerdict: fb.verdict,
          probeSeverity: fb.severity,
        });
      }
    }
  }

  sortAlertsBySeverity(alerts);
  return alerts;
}
